adapters/zio-http/src/main/scala/caliban/ZHTTPAdapter.scala

```diff
@@ -0,0 +1,240 @@
+    skipValidation: Boolean = false,
+    enableIntrospection: Boolean = true,
+    keepAliveTime: Option[Duration] = None,
+    queryExecution: QueryExecution = QueryExecution.Parallel
+  ) =
+    HttpApp.responseM(
+      for {
+        ref <- Ref.make(Map.empty[String, Promise[Any, Unit]])
+      } yield Response.socket(
+        socketHandler(
+          ref,
+          interpreter,
+          skipValidation,
+          enableIntrospection,
+          keepAliveTime,
+          queryExecution
+        )
+      )
+    )
+
+  private def socketHandler[R <: Has[_], E](
+    subscriptions: Ref[Map[String, Promise[Any, Unit]]],
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean,
+    enableIntrospection: Boolean,
+    keepAliveTime: Option[Duration],
+    queryExecution: QueryExecution
+  ): SocketApp[R with Clock, E] = {
+    val routes = Socket.collect[WebSocketFrame] { case Text(text) =>
+      ZStream
+        .fromEffect(ZIO.fromEither(decode[GraphQLWSRequest](text)))
+        .collect({
+          case GraphQLWSRequest("connection_init", _, _)      => connectionAck ++ keepAlive(keepAliveTime)
+          case GraphQLWSRequest("connection_terminate", _, _) => close
+          case GraphQLWSRequest("start", id, payload)         =>
+            payload match {
+              case Some(req) =>
+                generateGraphQLResponse(
+                  req,
+                  id.getOrElse(""),
+                  interpreter,
+                  skipValidation,
+                  enableIntrospection,
+                  queryExecution,
+                  subscriptions
+                ).catchAll(toStreamError(id, _))
+              case None      => connectionError
+            }
+          case GraphQLWSRequest("stop", id, _)                =>
+            removeSubscription(id, subscriptions)
+              .flatMap(_ => ZStream.empty)
+
+        })
+        .flatten
+        .catchAll(_ => connectionError)
+    }
+
+    SocketApp.message(routes) ++ SocketApp.protocol(protocol)
+  }
+
+  private def generateGraphQLResponse[R, E](
+    payload: GraphQLRequest,
+    id: String,
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean,
+    enableIntrospection: Boolean,
+    queryExecution: QueryExecution,
+    subscriptions: Subscriptions
+  ): ZStream[R, E, Text] =
+    ZStream
+      .fromEffect(
+        interpreter
+          .executeRequest(payload, skipValidation, enableIntrospection, queryExecution)
+      )
+      .flatMap(res =>
+        res.data match {
+          case ObjectValue((fieldName, StreamValue(stream)) :: Nil) =>
+            trackSubscription(id, subscriptions).flatMap { p =>
+              stream.map(toResponse(id, fieldName, _, res.errors)).interruptWhen(p)
+            }
+          case other                                                =>
+            ZStream.succeed(toResponse(id, GraphQLResponse(other, res.errors)))
+        }
+      )
+      .catchAll { _ =>
```

Comment:
same here

Reply:
Fixed in ebb17f919b719cfb09cb43bb74fbdb40c070fbb3.